### PR TITLE
test(e2e): multi-svc-test add nlb and udp connection test

### PR DESCRIPTION
**Pull request by CaptainCarpensir**
Adds a test case to an existing e2e test which adds a UDP network load balancer and tests for success with UDP traffic.
Should be merged after #5316 

e2e test for #4767 

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the Apache 2.0 License.

### CODE REVIEW

File: e2e/multi-svc-app/multi_svc_app_test.go [hunk 1]
```diff
@@ -242,17 +245,18 @@ var _ = Describe("Multiple Service App", func() {
 
 				// Call each environment's endpoint and ensure it returns a 200
 				route := svc.Routes[0]
+				routeURL = strings.Split(route.URL, " ")[0]
```
Comment (Lou1415926) [inline e2e/multi-svc-app/multi_svc_app_test.go]: Can we have a comment here showing examples of `route` and `routeURL`?

File: e2e/multi-svc-app/multi_svc_app_test.go [hunk 2]
```diff
@@ -338,6 +342,51 @@ var _ = Describe("Multiple Service App", func() {
...
+			svcName := "front-end"
+			svc, svcShowErr := cli.SvcShow(&client.SvcShowRequest{
+				AppName: appName,
+				Name:    svcName,
+			})
+			Expect(svcShowErr).NotTo(HaveOccurred())
+			Expect(len(svc.Routes)).To(Equal(1))
+
+			route := svc.Routes[0]
+			addr := strings.Split(route.URL, " ")[2]
```
Comment (Lou1415926) [inline e2e/multi-svc-app/multi_svc_app_test.go]: same! a commented example would be nice!

**Review by Lou1415926 — COMMENTED** [review]
awesome!

File: e2e/multi-svc-app/multi_svc_app_test.go [hunk 3]
```diff
@@ -243,16 +246,20 @@ var _ = Describe("Multiple Service App", func() {
 				// Call each environment's endpoint and ensure it returns a 200
 				route := svc.Routes[0]
 				Expect(route.Environment).To(Equal("test"))
+
+				// route.URL for "front-end" is of the form `(alb url) or (nlb url), so we split to retrieve just one valid url`
+				routeURL = strings.Split(route.URL, " ")[0]
```
Comment (iamhopaul123) [inline e2e/multi-svc-app/multi_svc_app_test.go]: Maybe we can do something safer here: do the split and the expect the result length to be at least 1 and then get the first element.

File: e2e/multi-svc-app/multi_svc_app_test.go [hunk 4]
```diff
@@ -243,16 +246,20 @@ var _ = Describe("Multiple Service App", func() {
 				// Call each environment's endpoint and ensure it returns a 200
 				route := svc.Routes[0]
 				Expect(route.Environment).To(Equal("test"))
+
+				// route.URL for "front-end" is of the form `(alb url) or (nlb url), so we split to retrieve just one valid url`
```
Comment (iamhopaul123) [inline e2e/multi-svc-app/multi_svc_app_test.go]: Can you give a real example for what `route.URL` will be like? Do we always get the ALB url?

File: e2e/multi-svc-app/front-end/main.go [hunk 5]
```diff
@@ -128,7 +130,30 @@ func PutEFSCheck(w http.ResponseWriter, req *http.Request, ps httprouter.Params)
 	w.WriteHeader(http.StatusOK)
 }
 
-func main() {
+func handleUDPTraffic(wg *sync.WaitGroup) {
+	defer wg.Done()
+
+	udpServer, err := net.ListenUDP("udp", &net.UDPAddr{Port: 8080})
```
Comment (iamhopaul123) [inline e2e/multi-svc-app/front-end/main.go:136]: Do we need to do `defer udpServer.Close()`?

File: e2e/multi-svc-app/front-end/main.go [hunk 6]
```diff
@@ -128,7 +130,30 @@ func PutEFSCheck(w http.ResponseWriter, req *http.Request, ps httprouter.Params)
...
+	if err != nil {
+		log.Fatal(err)
+	}
+	log.Println("Listening on port 8080...")
+
+	for {
+		buffer := make([]byte, 1024)
```
Comment (iamhopaul123) [inline e2e/multi-svc-app/front-end/main.go]: I feel like we shouldn't init this read buf within each loop. Instead, we should probably only init a print buf when the read succeeds and clear the read buf.

File: e2e/multi-svc-app/front-end/main.go [hunk 7]
```diff
@@ -140,3 +165,12 @@ func main() {
...
 	log.Fatal(http.ListenAndServe(":80", router))
 }
+
+func main() {
+	var wg sync.WaitGroup
+	wg.Add(2)
+	go handleUDPTraffic(&wg)
+	go handleHTTPTraffic(&wg)
+	wg.Wait()
+	log.Println("Finished listening")
```
Comment (iamhopaul123) [inline e2e/multi-svc-app/front-end/main.go:176]: Just want to double check: theoretically this should never happen?

Comment (CaptainCarpensir) [reply e2e/multi-svc-app/front-end/main.go:176]: Yep

File: e2e/multi-svc-app/multi_svc_app_test.go (hunk 4 above)
Comment (CaptainCarpensir) [reply e2e/multi-svc-app/multi_svc_app_test.go]: From testing yeah, we always get both the ALB and NLB url

File: e2e/multi-svc-app/multi_svc_app_test.go [hunk 8]
```diff
@@ -243,16 +246,23 @@ var _ = Describe("Multiple Service App", func() {
 				// Call each environment's endpoint and ensure it returns a 200
 				route := svc.Routes[0]
 				Expect(route.Environment).To(Equal("test"))
+
+				// route.URL is of the form `example-alb.elb.us-west-2.amazonaws.com or example-nlb.elb.us-west-2.amazonaws.com, so we split to retrieve just one valid url`
+				routeURLs := strings.Split(route.URL, " ")
+				Expect(len(svc.Routes)).To(BeNumerically(">", 1))
```
Comment (Lou1415926) [inline e2e/multi-svc-app/multi_svc_app_test.go]: Will this really be `true`? I might have missed something - it seems like we are only deploying to `test` environment - so there will be only one route in `svc.Routes`. 

Or maybe you meant `Expect(len(routeURLs)).To(BeNumerically(">", 1))`?

**Review by Lou1415926 — APPROVED** [review]

File: e2e/multi-svc-app/multi_svc_app_test.go (hunk 8 above)
Comment (CaptainCarpensir) [reply e2e/multi-svc-app/multi_svc_app_test.go]: OOPS good catch thank you, I did mean `Expect(len(routeURLs)).To(BeNumerically(">", 1)`

File: e2e/multi-svc-app/multi_svc_app_test.go [hunk 9]
```diff
@@ -338,6 +352,56 @@ var _ = Describe("Multiple Service App", func() {
...
+					Since:   "1h",
+				})
+				return svcLogs, svcLogsErr
+			}, "60s", "10s").ShouldNot(BeEmpty())
+
+			var svcLogMessages []string
+			for _, logLine := range svcLogs {
+				svcLogMessages = append(svcLogMessages, logLine.Message)
+			}
+			Expect(svcLogMessages).To(ContainElement(ContainSubstring("Received UDP message: test message")))
```
Comment (dannyrandall) [inline e2e/multi-svc-app/multi_svc_app_test.go]: should we parse the messages in the `Eventually()` above?

**Review by dannyrandall — COMMENTED** [review]
looks good, just one question!

File: e2e/multi-svc-app/multi_svc_app_test.go (hunk 9 above)
Comment (CaptainCarpensir) [reply e2e/multi-svc-app/multi_svc_app_test.go]: I see your point here. I guess the concern is that there are log entries that aren't from the UDP traffic but cause the `Eventually()` to return, and then we miss the UDP packet if for some reason it's slow to arrive or be logged. I find this possibility insanely unlikely given the timescale we're checking on, but I guess it's still possible 👍

**Review by iamhopaul123 — APPROVED** [review]

**Pull request merged**